### ISSUE TITLE
Design Proposal: Allow overzoom for sparse pyramids tiles

### ISSUE THREAD

**Issue by etiennejourdier** [maplibre/maplibre-style-spec#938]
# Design Proposal: Allow overzoom for sparse pyramid tiles

## Motivation: help deploy light pmtiles

Vector tiles are a great opportunity for developers to host their own tile server rather than relying on a third-party service. However a full planet pmtiles in OpenMapTiles schema is more than 70 Go, which is too much for small-scale web hosting.

Hopefully most maps do not require full detail of the whole planet, but the whole planet only up to an intermediate zoom (let's say z10) to get a nice map, then full details (let's say z11-z14) only in an area of interest (a city, a state, a country, ...).

It is possible in pmtiles to store such sparse pyramid tiles, so I was expecting this behavior: 
![Image](https://github.com/user-attachments/assets/95877b6c-019d-4673-9fda-e4d8673f59cf)
- in the area of interest (right side): display the available z11-z14 tiles
- in the rest of the world (left side): overzoom in the last available level (z10)

But the acual maplibre behavior is this one, with nothing visible in the rest of the world
![Image](https://github.com/user-attachments/assets/2615c921-e730-4bd5-af18-a6a7e2fbe883)

This behavior is perfectly normal once you have noticed that the tile server logically responds an empty tile, so maplibre logically displays a transparent tile.

This behaviour is perfect when 2 sources have been declared : you want the user to see the 2nd source below when the 1st one is not available. But it is not suited for 1 sparse pmtiles source for which the "show me what's below" behavior would be to display the last available level.

MapLibre is able to overzoom the last available level when it is the maxzoom level and also when the tile server responds with a 404 error instead of an empty tile (this is the way I used to prepare the 1st image above)

## Proposed Change

I propose to allow the user to choose the expected behavior when the tile server responds an empty tile:
- the default behavior would be the actual one: display nothing (for people with multiple sources)
- a new available behavior would be to overzoom in the last available level (for people with 1 source), as is the case today when the tile server responds an error

This could be implemented in maplibre-style-spec during source definition, for example with a new parameter _overzoomEmptyTile: true/false_ or _emptyTileHandling: 'overzoom'/'transparent'_  that allows the user to choose the desired behavior

```
sources: {
  'openmaptiles': {
     type: 'vector',
     url: `pmtiles://${PMTILES_URL}`,
     overzoomEmptyTile: true,
  }
}
```
## API Modifications

I don't know what to write in this section.

## Migration Plan and Compatibility

I don't expect any compatibility issues, as the default behaviour would still be to display a transparent tile.

## Rejected Alternatives

Several people (in [this discussion on pmtiles repo](https://github.com/protomaps/PMTiles/discussions/321) and [this one on maplibre GL JS repo](https://github.com/maplibre/maplibre-gl-js/discussions/5126)) proposed to prepare 2 separate files and declare these as 2 sources. This solution is not suitable because:
1) 2 sources make 2 calls whereas only 1 is used
1) to apply the same style to both file sources, you have to duplicate the entire style definition. And for styles with a lot of layers (based on OpenMapTiles schema for example), it makes a lot of duplicated content.
2) more problematic, since the z1-z10 file is overzoomed everywhere, you get in the area of interest a strange superposition of both sources:

![Image](https://github.com/user-attachments/assets/ce2158eb-f33c-4252-bd3c-46a462ade40f) ![Image](https://github.com/user-attachments/assets/4d01de78-3349-4c3e-85e3-1fb6e4ea07cb)

This proposal will allow a single-file solution that can be easily deployed by people with few map-coding skills, on small-scale servers.

**Comment by prusswan** [comment]
I recognize this as an issue of variable effective maxzoom in a single tileset, which is something not accounted for by the vector tile specs (if a tileset declares maxzoom=n it *should* have the tiles at that level - whether the tile server respond with an error or blank tile should not matter). While I don't expect libraries to handle this, if a solution can be found for maplibre JS, then it can possibly be applied for other client libraries as well. In cases where you are using/combining external tile sources that you cannot modify, client side fixes would be necessary anyway.

**Comment by HarelM** [comment]
Can you put it in a table? I don't want to interpret the sentence wrong...
I agree the behavior should be consistent across web and native.

**Comment by xabbu42** [comment]
sure

For vector sources:

|   Server response |     204 |     404 |
| ------------------ | ------ | ------ |
| maplibre-gl-js       | empty | empty |
| maplibre-native    | empty | empty |

For raster sources:

|   Server response |     204      |          404 |
| ------------------ | ------ | ------ |
| maplibre-gl-js       | empty      | overscale |
| maplibre-native    | overscale | overscale |

empty: rendering an empty tile, i.e. a vector tile without any geometry or a transparent image
overscale: falling back to a tile with a lower zoom level and scaling it to the higher zoom level

**Comment by xabbu42** [comment]
actually I'm not sure I really tested the vector 404 case, will have to redo that. it seems strange to me that there is no fallback in that case...

**Comment by etiennejourdier** [comment]
I agree, in my case (vector tiles, maplibre-GL-JS, 404 error) I observed overscale behavior.

**Comment by xabbu42** [comment]
Just checked the 404 vector tiles case again with maplibre-gl-js 4.7.1 and maplibre-native glfw-app on mac. I don't see any overscaling at all.  So the tables seem to be correct as written.

**Comment by etiennejourdier** [comment]
This is really strange, I've deployed a sparse pyramid pmtiles on my website for 1 month, with pmtiles protocol throwing 404 errors (using new option errorOnMissingTile: true, see [here](https://github.com/protomaps/PMTiles/discussions/321)), and overscaling is working perfectly well on maplibre-gl-js.
You can see here the border between the 2 areas:
https://playguide.eu/app/playground/1738f899-d174-4be0-9987-03231b60645a
- south display z14 tiles 
- north display overscaled z12 tiles
- console display 404 errors

![Image](https://github.com/user-attachments/assets/1bb2b4f4-f703-40a3-a03a-39dbe2be47bc)

Do we use different versions of maplibre-gl-js?

**Comment by xabbu42** [comment]
I don't see any 404 errors in console or the network tab on your example. As far as I understand pmtiles uses range queries so thats a different case than the ones I tested. It is relevant though that the behavior for missing tiles is different when using pmtiles and range queries than with a full GET requests per tile.

**Comment by etiennejourdier** [comment]
Don't you see these errors in the console ? At least after scrolling the map ?

![Image](https://github.com/user-attachments/assets/31fb396c-3bd4-4899-83af-d6f596ccf0f8)

Maybe my understanding is wrong and they are not exactly "404" errors, but for sure they are errors thrown [here by pmtiles.js](https://github.com/protomaps/PMTiles/blob/d4d97ad7a9c01f8a87f1e2e2e0cb3a177c933ca8/js/src/adapters.ts#L267) and caught by maplibre-gl-js, which triggers overscaling.

With the default option errorOnMissingTile: false, [pmtiles.js returns a blank array](https://github.com/protomaps/PMTiles/blob/d4d97ad7a9c01f8a87f1e2e2e0cb3a177c933ca8/js/src/adapters.ts#L269), and maplibre-gl-js display an empty tile.

**Comment by xabbu42** [comment]
I see those errors and they are not 404 errors so there is no conflict between our observations, just two different data points. Good to know that the default for pmtiles is the same as in my case and you enabled the difference with an option.

Can you test what the default is with a sparse raster pmtile by any chance to cover all cases?
